### PR TITLE
API: Backend

### PR DESCRIPTION
**Pull request by guydc**
**What this PR does / why we need it**:

**This is a preliminary API proposal for discussion purposes only.** The new `Backend` API allows users to reference UDS, FQDN and IP backends with an alternative to the K8s `Service` resources. See design doc for more details. 

**Which issue(s) this PR fixes**:
Relates to #2997, #36

### CODE REVIEW

**Comment by zirain** [comment]
so these's something like `ServiceEntry` in `Istio`?

File: site/content/en/latest/contributions/design/backend.md
Comment (arkodg) [inline site/content/en/latest/contributions/design/backend.md:111]: -1 on attaching BTS to a Backend, we technically have 4 levels of implicit defaults and overrides (`gateway`, `;listener`, `route` and `route rule` soon), adding more will further increase complexity

File: site/content/en/latest/contributions/design/backend.md
Comment (arkodg) [inline site/content/en/latest/contributions/design/backend.md:131]: is there a valid use to supports UDS addresses in HTTPRoute ? I can't think of any

Comment (guydc) [reply site/content/en/latest/contributions/design/backend.md:131]: If the gateway is deployed in a very coupled manner with the backend, it could make sense. For example, if both the backend(s) and gateway are daemonsets, they may communicate over a host-level UDS. 

Proxies do support this in general: [nginx](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#upstream), [haproxy](https://www.haproxy.com/documentation/haproxy-configuration-manual/latest/#4.2-server)

File: site/content/en/latest/contributions/design/backend.md
Comment (guydc) [reply site/content/en/latest/contributions/design/backend.md:111]: How is that different from supporting this for services, though? In terms of defaults/overrides, doesn't this introduce the same amount of layers that exists today?

**Comment by guydc** [comment]
@zirain, @Xunzhuo - There are some difference from `ServiceEntry`. 
- `ServiceEntry` doesn't support Unix Domain Sockets
- You can route from `VirtualService` to external services by just specifying their address. You don't need a `ServiceEntry` for that, but a `ServiceEntry` allows you to configure some setting related to DNS resolution, (m)TLS, etc. In EG, this resource (or `Service` + FQDN `Endpoint`) is mandatory for external routing.

File: site/content/en/latest/contributions/design/backend.md
Comment (guydc) [reply site/content/en/latest/contributions/design/backend.md:131]: As discussed in the community meeting, we can choose to reject UDS for xRoute BackendRefs. The External Service use case is still valuable here.

File: api/v1alpha1/backend_types.go
Comment (zhaohuabing) [inline api/v1alpha1/backend_types.go:32]: Should "WS" be a standalone ApplicationProtocolType?

File: api/v1alpha1/backend_types.go
Comment (zhaohuabing) [inline api/v1alpha1/backend_types.go:70]: Should follow [IANA standard service names](https://www.iana.org/assignments/service-names).

File: api/v1alpha1/backend_types.go
Comment (zhaohuabing) [inline api/v1alpha1/backend_types.go:21]: Address can be layer2/layer3/layer4/layer7, probably just name it as Protocol?

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:32]: I tried to keep this in-sync with https://gateway-api.sigs.k8s.io/geps/gep-1911/#supporting-protocols

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:70]: I think that eventually, we use this indication to create cluster settings that are protocol-specific. Envoy supports a closed list of protocols: HTTPx, TLS, TCP and some [other protos](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/other_protocols). So, I do think that it makes sense to limit this list to supported protocols only (both in envoy and envoy gateway).

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:95]:
```suggestion
	// +kubebuilder:validation:MinItems=1
```

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:61]: is name needed ?

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:62]: should we add a top level `Type` - FQDN, IPv4, IPv6, UDS ?

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:61]: I originally intended to use this to support named-section attachment of policies. However, since named-section attachment is not generally supported for xRoutes (only service ports are supported as a section), I think that we can decide that this is out of scope for now. We can add names later on to support this if needed.

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:62]: UDS is distinguished from other types with a dedicated struct, since it doesn't follow the `host:port` pattern anyway. We can add IPv4/IPv6/FQDN type field. The value that I see for such a type is simplification of validation (only need to check against the expected type). However, it creates a more verbose API. I'm OK with both. WDYT?

File: api/v1alpha1/backend_types.go
Comment (zirain) [inline api/v1alpha1/backend_types.go:29]: can we support IPv4 and IPv6 at the same time?

Comment (guydc) [reply api/v1alpha1/backend_types.go:29]: I couldn't find anything in the envoy docs to suggest that it's not possible. Judging by this example: https://istio.io/latest/blog/2023/experimental-dual-stack/, istio uses both IPv4 and IPv6 endpoints in the same cluster for dualstack support, as seen here:

```
ENDPOINT                 STATUS      OUTLIER CHECK     CLUSTER
[...]
10.244.0.26:9000         HEALTHY     OK                outbound|9000||tcp-echo.dual-stack.svc.cluster.local
fd00:10:244::1a:9000     HEALTHY     OK                outbound|9000||tcp-echo.dual-stack.svc.cluster.local
[...]
```

Having said that, I think that we can avoid IPv6 support in this API for the time being, and maybe address this at a later stage, together with #184.

File: api/v1alpha1/backend_types.go
Comment (zhaohuabing) [inline api/v1alpha1/backend_types.go:102]: Nit: uint32

Comment (guydc) [reply api/v1alpha1/backend_types.go:105]: In most of our APIs, we use int32 and kubebuilder min/max validations for portys. I think that this is because the CRDs are actually generated as int32, even when the go type is uint32, which will delay the validation to the processing phase. So, I kept it int32 but added a kubebuilder validation.

**Comment by zhaohuabing** [comment]
LGTM. Thanks for adding this valuable API. It unlocks the full potential of EG to many more use cases. 

Defer to other @envoyproxy/gateway-maintainers

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:110]: can we do w/o this field in the init PR and rely on info from the `xRoute` for the protocol ?

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:131]: can we do w/o this field in the init PR and rely on info from the xRoute for the protocol ?

File: api/v1alpha1/backend_types.go
Comment (zhaohuabing) [reply api/v1alpha1/backend_types.go:110]: Would this result in protocol mismatching if the protocol of `xRoute` is not the same as the `Backend`?

File: api/v1alpha1/backend_types.go
Comment (arkodg) [reply api/v1alpha1/backend_types.go:82]: thoughts on ?
```
	FQDN *SocketAddress `json:"fqdn,omitempty"`
	
	IP *SocketAddress `json:"ip,omitempty"`

	// Unix defines the unix domain socket path
	//
	// +optional
	Unix *UnixDomainSocketAddress `json:"unix,omitempty"`
```

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:125]: its very wordy rn, `address` shows up in 3 layers
```
backendAddresses:
- socketAddress:
     address: 1.2.3.4
     port: 80
```

File: api/v1alpha1/backend_types.go
Comment (arkodg) [reply api/v1alpha1/backend_types.go:110]: that depends on the backend server type, so app dev should be using the right Route type imo

File: site/content/en/contributions/design/backend.md
Comment (arkodg) [inline site/content/en/contributions/design/backend.md:111]: for xRoute lets also disable `localhost` and `127.0.0.1`

File: site/content/en/contributions/design/backend.md
Comment (arkodg) [inline site/content/en/contributions/design/backend.md:139]: I'd vote to make this opt in at the `EnvoyGateway` level by an admin to avoid
https://github.com/advisories/GHSA-vw47-mr44-3jf9

File: api/v1alpha1/backend_types.go
Comment (zhaohuabing) [reply api/v1alpha1/backend_types.go:110]: We can add validation to check if there's a protocol mismatching if the Backend API also has a Protocol field and reflect it to the xRoute status. Otherwise, the mismatching leaks to runtime, making it harder for debugging.

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:131]: I think that it's not currently possible to infer if HTTP1 or HTTP2 should be used upstream in HTTP routes. So, we need something similar to K8s service port application protocols.

Comment (arkodg) [reply api/v1alpha1/backend_types.go:131]: suggest making it a list if we want to add it in the first iteration
relates to https://github.com/kubernetes/enhancements/issues/3726

reg appProtocol naming, can we reuse anything from upstream K8s https://github.com/kubernetes/enhancements/pull/3727 ?

File: site/content/en/contributions/design/backend.md
Comment (guydc) [reply site/content/en/contributions/design/backend.md:111]: I have no objection to doing this. However, one can easily bypass this check by using a FQDN that resolves to 127.0.0.1. This is also true for services with FQDN endpoints. 

The real solution here is to use something like a network policy to block this sort of egress traffic from the proxy pod. However, network policies are not supported in every K8s distribution (e.g. kind). 

I think that that covering this in a security doc is probably the best approach. 

WDYT?

Comment (arkodg) [reply site/content/en/contributions/design/backend.md:111]: for services with FQDN endpoints / endpointslices , the service is owned by the app dev (lower priviledge) and the endpointslice is owned by the admin (higher priviledge) , so if the admin wants to muck around, we can let them, but this is more around not allowing app devs to be able to access the proxy endpoint directly . I'd try to avoid as much as we can, and document what we can't in a separate doc

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:131]: > suggest making it a list

Can you explain why we need a list based on the KEP?

File: api/v1alpha1/backend_types.go
Comment (arkodg) [reply api/v1alpha1/backend_types.go:110]: yeah my vote would be to capture this as a issue and tackle once the need arises

File: api/v1alpha1/backend_types.go
Comment (arkodg) [reply api/v1alpha1/backend_types.go:131]: +1 to naming suggestion

Comment (arkodg) [reply api/v1alpha1/backend_types.go:131]: also prefer `appProtocols` over `applicationProocols`

**Comment by zirain** [comment]
will this help https://gateway.envoyproxy.io/v1.0.1/tasks/traffic/routing-outside-kubernetes/ without using service/endpointslice?

**Comment by arkodg** [comment]
> will this help https://gateway.envoyproxy.io/v1.0.1/tasks/traffic/routing-outside-kubernetes/ without using service/endpointslice?

yes this API is a better approach to do non k8s routing

File: site/content/en/contributions/design/backend.md
Comment (guydc) [reply site/content/en/contributions/design/backend.md]: Updated the design doc to reflect this. Will implement the validation in the translator, as it's specific to xRoutes and not all instances of `Backend`.

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go]: done

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go]: Agree. When ahead and created dedicated structs to make validation a bit easier. For a yaml perspective, FQDN/IP look the same.

File: api/v1alpha1/backend_types.go
Comment (arkodg) [reply api/v1alpha1/backend_types.go]: could be simplified to
```
spec:
  endpoints:
  - ipv4:
      address: 1.2.3.4
      port: 80   
```

```
spec:
  endpoints:
  - ipv6:
      address: fe80::42:acff:fe11:2
      port: 80   
```

```
spec:
  endpoints:
  - unix:
       path: /var/run/blah.sock
```

**Comment by zirain** [comment]
> > will this help https://gateway.envoyproxy.io/v1.0.1/tasks/traffic/routing-outside-kubernetes/ without using service/endpointslice?
> 
> yes this API is a better approach to do non k8s routing

@guydc is this scenario in your plan?

**Comment by guydc** [comment]
> > > will this help https://gateway.envoyproxy.io/v1.0.1/tasks/traffic/routing-outside-kubernetes/ without using service/endpointslice?
> > 
> > 
> > yes this API is a better approach to do non k8s routing
> 
> @guydc is this scenario in your plan?

I'm targeting two scenarios:
- Simplify routing to out-of-cluster endpoints (for xRoutes and other backendRefs), as you mention 
- Support unix domain socket communication for reduced latency to extension servers (Ext Auth, Ext Proc, RL, Telemetry sinks.. ) 

I don't intend to implement support for `Backend` everywhere immediately. I'll start out with HTTPRoutes and ExtProc support for `Backend` referencing. I hope that that other use cases will be implemented gradually based on requirement from users.

File: site/content/en/latest/contributions/design/backend.md
Comment (guydc) [reply site/content/en/latest/contributions/design/backend.md]: as discussed, this was a misunderstanding. BackendTrafficPolicy will not attach to backend resources.

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:65]: should we split this up into `IPv4` and `IPv6` or leave it like this and infer internally using a lib ?
cc @envoyproxy/gateway-maintainers

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:122]: imo `backend` prefix is not needed, the config looks like
```
backend:
  spec:
    backendAddresses
```

this could be simplified to
```
backend:
  spec:
    addresses:

Comment (guydc) [reply api/v1alpha1/backend_types.go:122]: Sorry, missed your comment. I'll rename to endpoints as you suggested.

Comment (arkodg) [reply api/v1alpha1/backend_types.go:122]: tbh I dont know which one is better - `endpoints` or `addresses`

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:82]: `Host` or `Address` ?

Comment (guydc) [reply api/v1alpha1/backend_types.go]: With endpoints as a top-level, we can have address here, which I feel is more accurate. Updated it now,

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go]: ok, makes sense. Pluralized it.

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go]: I think that your suggestion is closer to both K8s Endpoint and Envoy Endpoint APIs, where endpoints have an address (or addresses). So, I think that it should be pretty intuitive to EG users.

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:67]:
```suggestion
	IPv4 *IPv4Endpoint `json:"ipv4,omitempty"`
```

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:131]:
```suggestion
	Endpoints []BackendEndpoint `json:"endpoints,omitempty"`
```

File: api/v1alpha1/backend_types.go
Comment (arkodg) [inline api/v1alpha1/backend_types.go:104]: should this be `Host` or `Address`

Comment (guydc) [reply api/v1alpha1/backend_types.go:104]: Envoy Endpoints contain an address ([IP, FQDN](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-socketaddress) or [UDS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-pipe)). The term envoy uses for hostname/ip is "address". 

```sh
Endpoint:
{
  "address": {...},
[...]
}

Socket Address:
{
  "protocol": ...,
  "address": ...,
  "port_value": ...,
[...]
}
```

Kubernetes Endpoints contain an address (IP, FQDN), and refer to hostname/ip as "address". 
```sh
apiVersion: discovery.k8s.io/v1
kind: EndpointSlice
metadata:
  name: example-abc
addressType: IPv4
endpoints:
  - addresses:
      - "10.1.2.3"
```

So, I think that we can stick to that terminology.

File: site/content/en/contributions/design/backend.md
Comment (shawnh2) [inline site/content/en/contributions/design/backend.md:2]:
```suggestion
title: "Backend"
```

File: site/content/en/contributions/design/backend.md
Comment (shawnh2) [inline site/content/en/contributions/design/backend.md:151]: can we use the relevant path for this?

File: site/content/en/contributions/design/backend.md
Comment (shawnh2) [inline site/content/en/contributions/design/backend.md:82]:
```suggestion
          name: backend-svc
```

do you mean this?

File: site/content/en/contributions/design/backend.md
Comment (shawnh2) [inline site/content/en/contributions/design/backend.md:35]: curious about here: do you mean we can configure a route that forwards traffic to a Service `backend-svc`, a locl UDS, and `10.244.0.28:3000`?

File: api/v1alpha1/backend_types.go
Comment (shawnh2) [inline api/v1alpha1/backend_types.go:45]:
```suggestion
	// Spec defines the desired state of Backend.
```

File: api/v1alpha1/backend_types.go
Comment (shawnh2) [inline api/v1alpha1/backend_types.go:48]:
```suggestion
	// Status defines the current status of Backend.
```

File: api/v1alpha1/backend_types.go
Comment (shawnh2) [inline api/v1alpha1/backend_types.go:49]: can you also include some basic status type and conditions in this PR as well?

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go]: Discussed in the community meeting and agreed on hostname.

File: api/v1alpha1/backend_types.go
Comment (guydc) [reply api/v1alpha1/backend_types.go:49]: done

File: site/content/en/contributions/design/backend.md
Comment (guydc) [reply site/content/en/contributions/design/backend.md]: yes, thanks, fixed

**Review by arkodg — APPROVED** [review]
LGTM thanks ! looking forward to using this API

**Review by zhaohuabing — APPROVED** [review]
LGTM thanks!

**Pull request merged**
